### PR TITLE
Added CanReadType to InputFormatter and Added serializer cache

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #1974 @yishaigalatzer @kichalla

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerInputFormatter.cs [hunk 1]
```diff
@@ -19,6 +20,7 @@ namespace Microsoft.AspNet.Mvc.Xml
     public class XmlDataContractSerializerInputFormatter : InputFormatter
     {
         private DataContractSerializerSettings _serializerSettings;
+        private ConcurrentDictionary<Type, object> _serializerCache = new ConcurrentDictionary<Type, object>();
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerInputFormatter.cs]: @yishaigalatzer : should this be static? even though formatters by default are registered as single instance, users can make them per-request too.

Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerInputFormatter.cs]: Static is a no go. If the user wants to go per request, he is responsible for caching. If we really cared we will have to provide another singleton service just for this. Because at this point he is already customizing it

File: src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs [hunk 2]
```diff
@@ -183,11 +185,31 @@ public override Task WriteResponseBodyAsync([NotNull] OutputFormatterContext con
...
+                var dataContractSerializer = GetCachedSerializer(wrappingType);
                 dataContractSerializer.WriteObject(xmlWriter, obj);
             }
 
             return Task.FromResult(true);
         }
+
+        private DataContractSerializer GetCachedSerializer(Type type)
+        {
+            if (type == null)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: why this?

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlDataContractSerializerInputFormatterTest.cs [hunk 3]
```diff
@@ -485,5 +502,16 @@ private InputFormatterContext GetInputFormatterContext(byte[] contentBytes, Type
             httpContext.SetupGet(c => c.Request).Returns(request.Object);
             return httpContext.Object;
         }
+
+        private class TestXmlDataContractSerializerInputFormatter : XmlDataContractSerializerInputFormatter
+        {
+            public int calledCount = 0;
```
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlDataContractSerializerInputFormatterTest.cs]: nit: rename this to something like `CreateSerializerCallCount` (`formatter.CreateSerializerCallCount` feels better than `formatter.callCount`)

File: src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs (hunk 2 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: `TryGetValue` produces an exception when `null` is passed. We have some tests that pass null to `CanRead` and expect `false`.
Eg. https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNet.Mvc.Xml.Test/XmlSerializerInputFormatterTest.cs#L50

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlDataContractSerializerInputFormatterTest.cs [hunk 4]
```diff
@@ -80,6 +79,24 @@ public void CanRead_ReturnsTrueForAnySupportedContentType(string requestContentT
...
+        {
+            // Arrange
+            var input = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+                "<DummyClass><SampleInt>10</SampleInt></DummyClass>";
+            var formatter = new TestXmlDataContractSerializerInputFormatter();
+            var contentBytes = Encoding.UTF8.GetBytes(input);
+            var context = GetInputFormatterContext(contentBytes, typeof(DummyClass));
+
+            // Act
+            formatter.CanRead(context);
```
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlDataContractSerializerInputFormatterTest.cs]: I see that you are trying to invoke in such a way that the CreateSerializer is potentially invoked more than once...but this intent is not clear here...can you call `await formatter.ReadAsync(...)` two times instead (by passing in newed up context)...

File: src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs (hunk 2 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: Did you mean a different test as in the link you provided, its about request content type and not the model type

File: src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerInputFormatter.cs [hunk 5]
```diff
@@ -142,7 +150,32 @@ protected virtual Type GetSerializableType([NotNull] Type declaredType)
...
+            }
+            catch (Exception)
+            {
+                // We do not surface the caught exception because if CanRead returns
+                // false, then this Formatter is not picked up at all.
+                return null;
+            }
+        }
+
+        private DataContractSerializer GetCachedSerializer(Type type)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerInputFormatter.cs]: A type's serializer could be cached 2 times (input and output formatters), but I prefer this as there is clean separation between input and output serializers instead of having a common cache...ex: probably there can be a scenario where a type's serializer has different settings for input than output...

File: src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs (hunk 2 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: also if a test is passing null for `Type` here, not sure how realistic it is..you should modify the test instead i think...

**Comment by kichalla** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs (hunk 2 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: Sorry. https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNet.Mvc.Xml.Test/XmlSerializerOutputFormatterTest.cs#L307

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: But it also changes the behavior. What is the expected behavior? Should `CanRead` return `false` for `null` or should it throw?

Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: This is in the output formatter...so its CanWriteType?...can you return `false` if ResolveType returns null?

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Xml/XmlDataContractSerializerOutputFormatter.cs]: Sorry again. I meant the CanWriteType. Yes, that can be done. :+1:

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by kichalla** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Xml/XmlSerializerOutputFormatter.cs [hunk 6]
```diff
@@ -159,11 +166,26 @@ public override Task WriteResponseBodyAsync([NotNull] OutputFormatterContext con
...
 
-                var xmlSerializer = CreateSerializer(wrappingType);
+                var xmlSerializer = GetCachedSerializer(wrappingType);
                 xmlSerializer.Serialize(xmlWriter, obj);
             }
 
             return Task.FromResult(true);
         }
+
+        private XmlSerializer GetCachedSerializer(Type type)
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Xml/XmlSerializerOutputFormatter.cs]: all these `GetCachedSerializer` need to be protected. They are breaking a creation of a different serializer per request (which the user can do). So if the user wants to create serializer per request he can override caching (or cache anyway he wants to). This also resolves @kichalla question about statics above :watch:

**Comment by yishaigalatzer** [comment]
:shipit:
